agent/consul/autopilot/autopilot.go

```diff
@@ -520,3 +528,27 @@ func IsPotentialVoter(suffrage raft.ServerSuffrage) bool {
 		return false
 	}
 }
+
+//For each member in serfWAN check to see if it
+//is failed - then check to see if it is a serfLAN
+//member, remove it.
+func (a *Autopilot) pruneSerfWAN(serfWAN *serf.Serf, serfLAN *serf.Serf) {
+	for _, memberWAN := range serfWAN.Members() {
```

Comment:
I have very limited context about what this code is doing, so totally ignore this if it seems inappropriate or not important!

This looks like it could potentially be an expensive operation with the 2 nested for loops. Maybe this isn't super important, but maybe we can make the algorithm more efficient? Perhaps by caching something from the first or second for loop, and then just having 2 side-by-side loops? Maybe first iterating through either the serfLAN or serfWAN members and caching our findings in a way that is easy to look up (like in a map)? (making the algorithm go from O(n^2) => O(2n)). 

Let me know if I’m not making sense.

Reply:
Hey -
It *was* super expensive (I whined about this function and it's complexity, as well as it's poor readability imo) - I wouldn't want to cache anything for Serf due to the possibility of harming our results. 
What this is s'posed to be doing is retrieving all the Serf members from WAN and LAN, comparing the results, and removing any failed members. The reason we did it like this was because we wanted to find any members in LAN after they have been removed, and if we find one then we need to remove it from WAN. We were trying to not do the calculation for Raft twice (where we make sure we can still maintain a consensus). 

Luckily the new implementation gets rid of this function entirely. :)